src/system/PredefinedProcedure.cpp

```diff
@@ -34,7 +53,14 @@ PredefinedProcedure::overload(const vector<pair<TypeNode *, bool>> &pairs, bool
     return types_.back().get();
 }
 
-ProcedureTypeNode *PredefinedProcedure::dispatch(vector<TypeNode *> actuals) const {
+ProcedureTypeNode *PredefinedProcedure::dispatch(vector<TypeNode *> actuals, TypeNode *typeType) const {
+    if (castIdx_ < 0) {
+        return nullptr;
+    }
+    if ((castIdx_ > 0) && typeType) {
+        auto signature = make_unique<ProcedureTypeNode>(EMPTY_POS, this->getIdentifier(), std::move(types_[0]->parameters()), types_[0]->hasVarArgs(), std::move(typeType));
+        return std::move(signature.get());
```

Comment:
This looks unsafe to me. The code creates a unique pointer to a signature and then returns a reference to this unique pointer. Since the unique pointer is deleted at the end of the block, the returned reference can be null. Also, I think the `std::move` is not required here.

Reply:
Used a new member field to keep pointer alive.
Not sure this is optimal.